### DISCUSSION TITLE
Does `self.log(..., on_epoch=True)` cost the same time as `torchmetrics.Accuracy` in distributed mode?

### DISCUSSION THREAD

**Discussion by marsggbo** [Lightning-AI/pytorch-lightning#9172] [DDP / multi-GPU / multi-node]
```python
class MyNet(pl.LightningModule):
    def __init__(self, ...):
        self.train_accuracy = Accuracy()

    def training_step(self, batch: Any, batch_idx: int):
        ...
        acc = self.train_accuracy(predictions, targets)

        self.log('train/acc', on_step=True, on_epoch=True)
```

In pl, there are two methods to calculate the accuracy of distributed mode. As shown in the above code, it seems that we calculate the accuracy twice, which wastes a lot of time. Therefore, I wonder how to save time in calculating the distributed metrics if we want to log and obtain the value of the metrics.

**Comment by carmocca** [comment]
You are right that this is duplicating the calculation. Instead, try logging the `Metric` directly:

```python
    def training_step(self, batch: Any, batch_idx: int):
        ...
        self.train_accuracy(predictions, targets)
        self.log('train/acc', self.train_accuracy, on_step=True, on_epoch=True)
```
